File: tests/docker_manager.py
"""
Docker container management for BDD testing framework.
"""
import docker
import logging
import os
import time
import yaml
from pathlib import Path
from typing import Dict, Optional, Any
from contextlib import contextmanager

logger = logging.getLogger(__name__)


class DockerManager:
    """Manages Docker containers for CLI testing scenarios."""
    
    def __init__(self, config_path: str = "tests/config.yaml"):
        self.config = self._load_config(config_path)
        self.client = docker.from_env()
        self.containers: Dict[str, docker.models.containers.Container] = {}
        self.images: Dict[str, str] = {}
        
    def _load_config(self, config_path: str) -> Dict[str, Any]:
        """Load configuration from YAML file."""
        try:
            with open(config_path, 'r') as f:
                return yaml.safe_load(f)
        except FileNotFoundError:
            logger.error(f"Config file not found: {config_path}")
            raise
            
    def build_environment_image(self, env_name: str) -> str:
        """Build Docker image for specified environment."""
        if env_name not in self.config['testing']['environments']:
            raise ValueError(f"Unknown environment: {env_name}")
            
        env_config = self.config['testing']['environments'][env_name]
        dockerfile_path = env_config['dockerfile']
        tag = env_config['tag']
        
        if not os.path.exists(dockerfile_path):
            raise FileNotFoundError(f"Dockerfile not found: {dockerfile_path}")
            
        logger.info(f"Building image {tag} from {dockerfile_path}")
        
        # Build context is the directory containing the Dockerfile
        build_context = str(Path(dockerfile_path).parent)
        dockerfile_name = Path(dockerfile_path).name
        
        try:
            image, logs = self.client.images.build(
                path=build_context,
                dockerfile=dockerfile_name,
                tag=tag,
                rm=True,
                timeout=self.config['testing']['docker']['build_timeout'],
                pull=True
            )
            
            # Log build output
            for log in logs:
                if 'stream' in log:
                    logger.debug(f"Build: {log['stream'].strip()}")
                    
            self.images[env_name] = tag
            logger.info(f"Successfully built image: {tag}")
            return tag
            
        except docker.errors.BuildError as e:
            logger.error(f"Failed to build image {tag}: {e}")
            raise
            
    def start_container(self, env_name: str, container_name: str = None) -> docker.models.containers.Container:
        """Start a container for the specified environment."""
        if env_name not in self.config['testing']['environments']:
            raise ValueError(f"Unknown environment: {env_name}")
            
        # Build image if not already built
        if env_name not in self.images:
            self.build_environment_image(env_name)
            
        tag = self.images[env_name]
        container_name = container_name or f"lintair-test-{env_name}-{int(time.time())}"
        
        # Mount the CLI binary into the container
        cli_binary_host = os.path.abspath("lintair")
        if not os.path.exists(cli_binary_host):
            raise FileNotFoundError(f"CLI binary not found: {cli_binary_host}")
            
        volumes = {
            cli_binary_host: {
                'bind': self.config['testing']['cli']['binary_path'],
                'mode': 'ro'
            },
            os.path.abspath("tests/test_files"): {
                'bind': '/test_files',
                'mode': 'rw'
            }
        }
        
        logger.info(f"Starting container {container_name} from image {tag}")
        
        try:
            container = self.client.containers.run(
                tag,
                name=container_name,
                detach=True,
                tty=True,
                stdin_open=True,
                volumes=volumes,
                network_mode=self.config['testing']['docker']['network_mode'],
                auto_remove=self.config['testing']['docker']['auto_remove'],
                working_dir="/test_files"
            )
            
            # Wait for container to be ready
            self._wait_for_container(container)
            
            self.containers[container_name] = container
            logger.info(f"Container {container_name} started successfully")
            return container
            
        except docker.errors.ContainerError as e:
            logger.error(f"Failed to start container {container_name}: {e}")
            raise
            
    def _wait_for_container(self, container: docker.models.containers.Container, timeout: int = 10):
        """Wait for container to be ready."""
        start_time = time.time()
        while time.time() - start_time < timeout:
            container.reload()
            if container.status == 'running':
                # Test if we can execute commands
                try:
                    result = container.exec_run("echo 'ready'", timeout=5)
                    if result.exit_code == 0:
                        return
                except:
                    pass
            time.sleep(0.5)
            
        raise TimeoutError(f"Container {container.name} not ready after {timeout}s")
        
    def execute_command(self, container_name: str, command: str, timeout: int = None) -> Dict[str, Any]:
        """Execute a command in the specified container."""
        if container_name not in self.containers:
            raise ValueError(f"Container not found: {container_name}")
            
        container = self.containers[container_name]
        timeout = timeout or self.config['testing']['cli']['default_timeout']
        
        logger.debug(f"Executing in {container_name}: {command}")
        
        try:
            result = container.exec_run(
                command,
                stdout=True,
                stderr=True,
                stdin=False,
                tty=False,
                privileged=False,
                user='',
                environment=None,
                workdir='/test_files',
                detach=False,
                stream=False,
                socket=False,
                demux=True,
                timeout=timeout
            )
            
            stdout = result.output[0].decode('utf-8') if result.output[0] else ""
            stderr = result.output[1].decode('utf-8') if result.output[1] else ""
            
            command_result = {
                'exit_code': result.exit_code,
                'stdout': stdout,
                'stderr': stderr,
                'command': command,
                'container': container_name
            }
            
            logger.debug(f"Command result: exit_code={result.exit_code}")
            if stdout:
                logger.debug(f"STDOUT: {stdout}")
            if stderr:
                logger.debug(f"STDERR: {stderr}")
                
            return command_result
            
        except Exception as e:
            logger.error(f"Failed to execute command in {container_name}: {e}")
            raise
            
    def get_container_logs(self, container_name: str) -> str:
        """Get logs from specified container."""
        if container_name not in self.containers:
            return ""
            
        container = self.containers[container_name]
        try:
            return container.logs().decode('utf-8')
        except Exception as e:
            logger.error(f"Failed to get logs from {container_name}: {e}")
            return ""
            
    def stop_container(self, container_name: str):
        """Stop and remove specified container."""
        if container_name not in self.containers:
            return
            
        container = self.containers[container_name]
        logger.info(f"Stopping container {container_name}")
        
        try:
            container.stop(timeout=10)
            if not self.config['testing']['docker']['auto_remove']:
                container.remove()
            del self.containers[container_name]
        except Exception as e:
            logger.error(f"Failed to stop container {container_name}: {e}")
            
    def cleanup_all_containers(self):
        """Stop and remove all managed containers."""
        for container_name in list(self.containers.keys()):
            self.stop_container(container_name)
            
    @contextmanager
    def container_context(self, env_name: str, container_name: str = None):
        """Context manager for container lifecycle."""
        container = None
        try:
            container = self.start_container(env_name, container_name)
            yield container
        finally:
            if container:
                self.stop_container(container.name)
                
    def list_environments(self) -> Dict[str, str]:
        """List available test environments."""
        return {
            name: config['description'] 
            for name, config in self.config['testing']['environments'].items()
        }

File: tests/docker_manager.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// DockerManager handles Docker operations for testing
type DockerManager struct {
	ctx context.Context
}

// ContainerContext holds information about a test container
type ContainerContext struct {
	ID           string
	Name         string
	Environment  string
	dockerFile   string
	tag          string
	manager      *DockerManager
	scenarioName string
}

// NewDockerManager creates a new Docker manager
func NewDockerManager() (*DockerManager, error) {
	// Check if Docker is available
	cmd := exec.Command("docker", "--version")
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("Docker not available: %w", err)
	}

	return &DockerManager{
		ctx: context.Background(),
	}, nil
}

// Close closes the Docker manager
func (dm *DockerManager) Close() error {
	return nil
}

// BuildImage builds a Docker image from a Dockerfile
func (dm *DockerManager) BuildImage(dockerFile, tag string) error {
	log.Printf("Building image %s from %s", tag, dockerFile)

	// Create a temporary directory for build context
	buildDir, err := os.MkdirTemp("", "lintair-build-*")
	if err != nil {
		return fmt.Errorf("failed to create build directory: %w", err)
	}
	defer os.RemoveAll(buildDir)

	// Copy Dockerfile to build directory
	dockerFileContent, err := os.ReadFile(dockerFile)
	if err != nil {
		return fmt.Errorf("failed to read Dockerfile: %w", err)
	}

	if err := os.WriteFile(filepath.Join(buildDir, "Dockerfile"), dockerFileContent, 0644); err != nil {
		return fmt.Errorf("failed to write Dockerfile: %w", err)
	}

	// Copy lintair binary to build directory
	// Try both relative and absolute paths
	binaryPath := "../lintair"
	if _, err := os.Stat(binaryPath); err != nil {
		// Try from current working directory
		binaryPath = "lintair"
	}
	if _, err := os.Stat(binaryPath); err == nil {
		binaryContent, err := os.ReadFile(binaryPath)
		if err != nil {
			return fmt.Errorf("failed to read binary: %w", err)
		}

		if err := os.WriteFile(filepath.Join(buildDir, "lintair"), binaryContent, 0755); err != nil {
			return fmt.Errorf("failed to write binary: %w", err)
		}
	} else {
		log.Printf("Warning: lintair binary not found at %s", binaryPath)
	}

	// Run docker build
	cmd := exec.Command("docker", "build", "-t", tag, buildDir)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("failed to build image: %w, output: %s", err, string(output))
	}

	log.Printf("Build output: %s", string(output))
	return nil
}

// CheckImageExists checks if a Docker image exists locally
func (dm *DockerManager) CheckImageExists(tag string) bool {
	cmd := exec.Command("docker", "images", "-q", tag)
	output, err := cmd.Output()
	return err == nil && len(strings.TrimSpace(string(output))) > 0
}

// NewContainerContext creates a new container context
func NewContainerContext(environment, dockerFile, tag string, manager *DockerManager) *ContainerContext {
	return &ContainerContext{
		Environment: environment,
		dockerFile:  dockerFile,
		tag:         tag,
		manager:     manager,
	}
}

// SetScenarioName sets the scenario name for container naming
func (cc *ContainerContext) SetScenarioName(scenarioName string) {
	cc.scenarioName = scenarioName
}

// StartContainer starts a new container for testing
func (cc *ContainerContext) StartContainer() error {
	// Build image if it doesn't exist locally
	if !cc.manager.CheckImageExists(cc.tag) {
		if err := cc.manager.BuildImage(cc.dockerFile, cc.tag); err != nil {
			return fmt.Errorf("failed to build image: %w", err)
		}
	}

	// Create and start container with scenario-based name
	scenarioKey := strings.ReplaceAll(strings.ToLower(cc.scenarioName), " ", "-")
	scenarioKey = strings.ReplaceAll(scenarioKey, ",", "")
	scenarioKey = strings.ReplaceAll(scenarioKey, "'", "")
	containerName := fmt.Sprintf("lintair-%s-%s", scenarioKey, cc.Environment)

	cmd := exec.Command("docker", "run", "-d", "--rm", "--name", containerName,
		"-w", "/tmp", cc.tag, "sleep", "300")
	output, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("failed to start container: %w, output: %s", err, string(output))
	}

	cc.ID = strings.TrimSpace(string(output))
	cc.Name = containerName

	log.Printf("Started container %s (%s)", cc.Name, cc.ID[:12])
	return nil
}

// StopContainer stops and removes the container
func (cc *ContainerContext) StopContainer() error {
	if cc.ID == "" {
		return nil
	}

	cmd := exec.Command("docker", "stop", cc.ID)
	if err := cmd.Run(); err != nil {
		log.Printf("Warning: failed to stop container %s: %v", cc.ID[:12], err)
	}

	log.Printf("Stopped container %s", cc.ID[:12])
	return nil
}

// CreateFile creates a file inside the container
func (cc *ContainerContext) CreateFile(filename, content string) error {
	// Create file using docker exec
	cmd := exec.Command("docker", "exec", cc.ID, "sh", "-c",
		fmt.Sprintf("cat > /tmp/%s", filename))
	cmd.Stdin = strings.NewReader(content)

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to create file %s: %w", filename, err)
	}

	log.Printf("Created file %s in container %s", filename, cc.ID[:12])
	return nil
}

// ExecuteCommand executes a command inside the container
func (cc *ContainerContext) ExecuteCommand(command string) (*CommandResult, error) {
	// Execute command using docker exec
	dockerCmd := exec.Command("docker", "exec", cc.ID, "sh", "-c", command)

	var stdout, stderr bytes.Buffer
	dockerCmd.Stdout = &stdout
	dockerCmd.Stderr = &stderr

	err := dockerCmd.Run()
	exitCode := 0
	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			exitCode = exitError.ExitCode()
		} else {
			exitCode = 1
		}
	}

	result := &CommandResult{
		Command:  command,
		ExitCode: exitCode,
		Stdout:   stdout.String(),
		Stderr:   stderr.String(),
	}

	log.Printf("Executed command: %s, exit code: %d", command, result.ExitCode)
	if result.Stdout != "" {
		log.Printf("Stdout: %s", result.Stdout)
	}
	if result.Stderr != "" {
		log.Printf("Stderr: %s", result.Stderr)
	}

	return result, nil
}

// CopyFileIntoContainer copies a file from the host into the container
func (cc *ContainerContext) CopyFileIntoContainer(sourcePath, destFilename string) error {
	// Use docker cp to copy file into container
	containerPath := fmt.Sprintf("%s:/tmp/%s", cc.ID, destFilename)
	cmd := exec.Command("docker", "cp", sourcePath, containerPath)

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to copy file %s to container: %w", sourcePath, err)
	}

	log.Printf("Copied file %s to container %s as %s", sourcePath, cc.ID[:12], destFilename)
	return nil
}

// VerifyLinterInstalled checks if a linter is installed in the container
func (cc *ContainerContext) VerifyLinterInstalled(linter string) bool {
	result, err := cc.ExecuteCommand(fmt.Sprintf("which %s", linter))
	if err != nil {
		return false
	}
	return result.ExitCode == 0
}

// CommandResult holds the result of a command execution
type CommandResult struct {
	Command  string
	ExitCode int
	Stdout   string
	Stderr   string
}
